nestjs/src/courses/team-distribution/distribute-students.service.ts

```diff
@@ -0,0 +1,189 @@
+      const team = teams.find(el => el.students.length < teamSize);
+      const teamDistributionStudent = shuffledStudents.pop();
+      if (teamDistributionStudent && team) {
+        team?.students.push(teamDistributionStudent.student);
+        distributedStudents.push(teamDistributionStudent);
+      }
+      capacity--;
+    }
+    await this.modifyTeams(
+      teams,
+      distributedStudents.map(el => ({ ...el, distributed: true })),
+    );
+  }
+
+  private async createInitialTeams(studentsCount: number, teamSize: number, teamDistributionId: number) {
+    const randomTeamsCount = Math.ceil(studentsCount / teamSize);
+    const teams: Pick<Team, 'name' | 'students' | 'description' | 'password' | 'teamDistributionId'>[] =
+      await Promise.all(
+        Array(randomTeamsCount)
+          .fill({})
+          .map(async (_, index) => {
+            const password = await this.teamService.generatePassword();
+            return {
+              name: `Random team #${index + 1}`,
+              students: [],
+              description: 'This team was created by random distribution.',
+              password,
+              teamDistributionId,
+            };
+          }),
+      );
+    return teams;
+  }
+
+  private async createRandomTeams(
+    teamDistributionStudents: TeamDistributionStudent[],
+    teamSize: number,
+    teamDistributionId: number,
+  ) {
+    const teams = await this.createInitialTeams(teamDistributionStudents.length, teamSize, teamDistributionId);
+
+    // Check if the number of teams is less than the team size
+    if (teams.length < teamSize) {
+      // If so, assign all students to the teams, making sure not to exceed the team size
+      const shuffledStudents = shuffle(teamDistributionStudents);
+      shuffledStudents.forEach(el => {
+        const team = teams.find(el => el.students.length < teamSize);
+        if (team) {
+          team?.students.push(el.student);
+        }
+      });
+    } else {
+      // If not, proceed with the draft-style distribution
+      const countDistributionRounds = Math.ceil(teamDistributionStudents.length / teamSize);
+      teams.forEach((team, i) => {
+        const students: Student[] = [];
+        for (let round = 1; round <= countDistributionRounds; round++) {
+          let draftPick;
+          if (round % 2 === 0) {
+            draftPick = round * teams.length - i;
+          } else {
+            draftPick = (round - 1) * teams.length + (i + 1);
+          }
+          if (teamDistributionStudents[draftPick - 1]) {
+            students.push(teamDistributionStudents[draftPick - 1]!.student);
+          }
+        }
+        team.students.push(...students);
+      });
+    }
+
+    // Save teams and teamDistributionStudent
+    await this.modifyTeams(
+      teams.map(t => ({ ...t, teamLeadId: t.students.sort((a, b) => a.rank - b.rank).at(0)?.id })),
+      teamDistributionStudents.map(s => ({ ...s, distributed: true })),
+    );
+  }
```

Comment:
suggest to break it into several steps for readability as it's important here.

Reply:
fixed